xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,743 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+	b.pickerUpdateCh.Put(s)
+}
+
+func (b *outlierDetectionBalancer) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	sc, err := b.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	scw := &subConnWrapper{
+		SubConn:    sc,
+		addresses:  addrs,
+		scUpdateCh: b.scUpdateCh,
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.scWrappers[sc] = scw
+	if len(addrs) != 1 {
+		return scw, nil
+	}
+	addrInfo, ok := b.addrs[addrs[0].Addr]
+	if !ok {
+		return scw, nil
+	}
+	addrInfo.sws = append(addrInfo.sws, scw)
+	atomic.StorePointer(&scw.addressInfo, unsafe.Pointer(addrInfo))
+	if !addrInfo.latestEjectionTimestamp.IsZero() {
+		scw.eject()
+	}
+	return scw, nil
+}
+
+func (b *outlierDetectionBalancer) RemoveSubConn(sc balancer.SubConn) {
+	scw, ok := sc.(*subConnWrapper)
+	if !ok { // Shouldn't happen
+		return
+	}
+	// Remove the wrapped SubConn from the parent Client Conn. We don't remove
+	// from map entry until we get a Shutdown state for the SubConn, as we need
+	// that data to forward that state down.
+	b.cc.RemoveSubConn(scw.SubConn)
+}
+
+// appendIfPresent appends the scw to the address, if the address is present in
+// the Outlier Detection balancers address map. Returns nil if not present, and
+// the map entry if present.
+func (b *outlierDetectionBalancer) appendIfPresent(addr string, scw *subConnWrapper) *addressInfo {
+	addrInfo, ok := b.addrs[addr]
+	if !ok {
+		return nil
+	}
+
+	addrInfo.sws = append(addrInfo.sws, scw)
+	atomic.StorePointer(&scw.addressInfo, unsafe.Pointer(addrInfo))
+	return addrInfo
+}
+
+// removeSubConnFromAddressesMapEntry removes the scw from its map entry if
+// present.
+func (b *outlierDetectionBalancer) removeSubConnFromAddressesMapEntry(scw *subConnWrapper) {
+	addrInfo := (*addressInfo)(atomic.LoadPointer(&scw.addressInfo))
+	if addrInfo == nil {
+		return
+	}
+	for i, sw := range addrInfo.sws {
+		if scw == sw {
+			addrInfo.sws = append(addrInfo.sws[:i], addrInfo.sws[i+1:]...)
+			break
+		}
+	}
+}
+
+func (b *outlierDetectionBalancer) UpdateAddresses(sc balancer.SubConn, addrs []resolver.Address) {
+	scw, ok := sc.(*subConnWrapper)
+	if !ok {
+		// Return, shouldn't happen if passed up scw
+		return
+	}
+
+	b.cc.UpdateAddresses(scw.SubConn, addrs)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Note that 0 addresses is a valid update/state for a SubConn to be in.
+	// This is correctly handled by this algorithm (handled as part of a non singular
+	// old address/new address).
+	switch {
+	case len(scw.addresses) == 1 && len(addrs) == 1: // single address to single address
+		// If the updated address is the same, then there is nothing to do
+		// past this point.
+		if scw.addresses[0].Addr == addrs[0].Addr {
+			return
+		}
+		b.removeSubConnFromAddressesMapEntry(scw)
+		addrInfo := b.appendIfPresent(addrs[0].Addr, scw)
+		if addrInfo == nil { // uneject unconditionally because could have come from an ejected address
+			scw.uneject()
```

Comment:
Maybe add a `break` here, so that the code in the `else` part can be less indented?